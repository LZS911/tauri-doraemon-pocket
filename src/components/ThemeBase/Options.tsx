import { useBoolean } from 'ahooks';
import { Popover } from 'antd';
import classNames from 'classnames';
import { IOptionsProps } from '.';
import usePropsValue from '../../hooks/usePropsValue';

const Options: <ValueType>(
  props: IOptionsProps<ValueType>
) => React.ReactElement = ({
  children,
  list,
  value,
  onChange,
  defaultValue,
  className = '',
  ...props
}) => {
  const [internalValue, setInternalValue] = usePropsValue({
    defaultValue,
    value,
    onChange,
  });
  const [open, { setFalse: closePopover, setTrue: openPopover }] = useBoolean();

  const onOpenChange = (open: boolean) => {
    if (open) {
      openPopover();
    } else {
      closePopover();
    }
  };

  return (
    <Popover
      {...props}
      className={`theme-options rounded-lg ${className}`}
      trigger="click"
      placement="bottom"
      onOpenChange={onOpenChange}
      open={open}
      content={list.map((v) => {
        return (
          <div
            onClick={() => {
              setInternalValue(v.value);
              v.onClick?.();
              closePopover();
            }}
            key={v.key}
            className={classNames(
              {
                ['options-selected-item bg-secondary hover:!bg-secondary']:
                  internalValue === v.value,
              },
              'w-48 cursor-pointer rounded-sm px-4 py-2 transition-[background] hover:bg-slate-200 dark:text-white dark:hover:bg-slate-500'
            )}
          >
            <span className="leading-6">{v.text}</span>
          </div>
        );
      })}
    >
      <div onClick={openPopover}>{children}</div>
    </Popover>
  );
};

export default Options;
