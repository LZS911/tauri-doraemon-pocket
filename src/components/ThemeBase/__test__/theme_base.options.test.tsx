import { fireEvent, render, screen } from '@testing-library/react';
import { IOptionsProps } from '..';
import Options from '../Options';

const onClickSpy1 = jest.fn();
const onClickSpy2 = jest.fn();

const list: IOptionsProps<string>['list'] = [
  {
    key: 'key1',
    value: 'value1',
    text: 'text1',
    onClick: onClickSpy1,
  },
  {
    key: 'key2',
    value: 'value2',
    text: 'text2',
    onClick: onClickSpy2,
  },
];

const children = <>children</>;

const getChildren = () => screen.getByText('children');

describe('test Options', () => {
  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });
  test('should match snapshot', () => {
    const { baseElement } = render(
      <Options list={list} defaultVisible={true} className="test">
        {children}
      </Options>
    );
    fireEvent.click(getChildren());

    expect(baseElement).toMatchSnapshot();
  });

  test('should be set default value when defaultValue is not empty', () => {
    render(
      <Options list={list} defaultValue={list[0].value}>
        {children}
      </Options>
    );

    fireEvent.click(getChildren());

    // expect(screen.getByText(list[0].text).parentElement).toHaveClass(
    //   'options-selected-item'
    // );
  });

  test('should execute onClick and onChange when clicking item', () => {
    const onChange = jest.fn();
    render(
      <Options list={list} defaultVisible={true} onChange={onChange}>
        {children}
      </Options>
    );
    fireEvent.click(getChildren());

    expect(onClickSpy1).toBeCalledTimes(0);
    expect(onClickSpy2).toBeCalledTimes(0);

    fireEvent.click(screen.getByText(list[0].text));
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toBeCalledWith(list[0].value);
    expect(onClickSpy1).toBeCalledTimes(1);

    fireEvent.click(screen.getByText(list[1].text));
    expect(onChange).toBeCalledTimes(2);
    expect(onChange).toBeCalledWith(list[1].value);
    expect(onClickSpy2).toBeCalledTimes(1);
  });
});
