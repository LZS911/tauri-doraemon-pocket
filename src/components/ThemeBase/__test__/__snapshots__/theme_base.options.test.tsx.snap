// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Options should match snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-popover-open"
    >
      children
    </div>
  </div>
  <div
    style="position: absolute; top: 0px; left: 0px; width: 100%;"
  >
    <div>
      <div
        class="ant-popover dark:bg-darkMode rounded-lg theme-options test theme-popover"
        style="opacity: 0;"
      >
        <div
          class="ant-popover-content"
        >
          <div
            class="ant-popover-arrow"
          >
            <span
              class="ant-popover-arrow-content"
            />
          </div>
          <div
            class="ant-popover-inner"
            role="tooltip"
          >
            <div
              class="ant-popover-inner-content"
            >
              <div
                class="w-48 cursor-pointer rounded-sm py-2 px-4 transition-[background] hover:bg-slate-200 dark:text-white dark:hover:bg-slate-500"
              >
                <span
                  class="leading-6"
                >
                  text1
                </span>
              </div>
              <div
                class="w-48 cursor-pointer rounded-sm py-2 px-4 transition-[background] hover:bg-slate-200 dark:text-white dark:hover:bg-slate-500"
              >
                <span
                  class="leading-6"
                >
                  text2
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
